### PR TITLE
Add design-docs folder and README.

### PR DESCRIPTION
**Pull request by murphyjacob4**
These are forked from the RFC instructions created by @zuiderkwast and @hpatro in https://github.com/valkey-io/valkey-rfc/pulls/1 and https://github.com/valkey-io/valkey-rfc/pulls/6

### CODE REVIEW

**Review by sarthakaggarwal97 — APPROVED** [review]

File: design-docs/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+   Changes to the design document can be committed alongside the code changes as
+   the mid-level design details are implemented.
+
+Each file has one of the following statuses:
+
+- **Proposed**, design has been proposed, and if merged, accepted.
+- **Partially Implemented**, design has been accepted, and some changes have
+  been implemented.
+- **Implemented**, design has been accepted, and all changes have been
+  implemented.
```
Comment (zuiderkwast) [inline design-docs/README.md]: This is the RFC workflow. It's quite far from the process I'd like to see for design docs.

I think a design docs should match the implementation and shipped at the same time, i.e. in the same PR as implementation. A design that's not implemented can stay as an open PR. No need for a status.

File: design-docs/TEMPLATE.md
Comment (zuiderkwast) [inline design-docs/TEMPLATE.md:1]: I don't think we should have a template. It was a mistake to add it in the RFC repo. Each design is different and don't have the same sections. Starting from an empty page, authors will naturally keep it more minimalistic and only write what they see relevant.

**Review by zuiderkwast — COMMENTED** [review]
I'm not sure if we should add this without adding any actual design. Are there any complete design doc in the RFC repo that we can move over?

File: design-docs/TEMPLATE.md
Comment (hpatro) [reply design-docs/TEMPLATE.md:1]: I think the downside of not having a template is that a new author in the project might overlook the various components which exists in the system.

**Review by hpatro — COMMENTED** [review]
With AI agents being a reality I think it makes sense to have design document along the main repository and keep it updated along with the code. Did you had any other thoughts in mind behind putting this up?

**Comment by hpatro** [comment]
Thank you @sarthakaggarwal97 for pointing me to the issue https://github.com/valkey-io/valkey/issues/3172

**Comment by murphyjacob4** [comment]
> With AI agents being a reality I think it makes sense to have design document along the main repository and keep it updated along with the code.

Yeah that was part of what we had discussed when a few weeks back in the Core Team sync. The goal would be that maintainers have those proposing substantial edits also edit the design documents

> Did you had any other thoughts in mind behind putting this up?

The primary other thinking was visibility for RFCs. The feedback was no one is reviewing RFCs. So putting them into the main repo should help get that visibility

File: design-docs/README.md (hunk 1 above)
Comment (murphyjacob4) [reply design-docs/README.md]: Should we force implementation and design at the same time? There may be cases where the design should get committed before the code, right? E.g. for large changes where we want alignment before code is being reviewed

Maybe we can say "you may send design and code in the same PR" but not force it. If you want to send just a design update that should be fine too

File: design-docs/TEMPLATE.md
Comment (murphyjacob4) [reply design-docs/TEMPLATE.md:1]: Right, most of the sections are optional so you can just delete them if they don't apply

**Comment by murphyjacob4** [comment]
> I'm not sure if we should add this without adding any actual design. Are there any complete design doc in the RFC repo that we can move over?

I can also move over JSON, Bloom, LDAP (our approved RFCs)

**Comment by murphyjacob4** [comment]
> I can also move over JSON, Bloom, LDAP (our approved RFCs)

Actually - should we hold off on these ones? I feel like their designs should ideally go into the module repo itself, not the core. Otherwise you end up making a lot of changes to the core repo when designs change

@zuiderkwast WDYT?

File: design-docs/TEMPLATE.md
Comment (zuiderkwast) [reply design-docs/TEMPLATE.md:1]: People don't often do that.

I think we'll see lot of boilerplate and duplicated information. All of the API docs belong in user-facing docs.

**Comment by zuiderkwast** [comment]
Why not start with the ASM design doc?

I would like some actual content first, rather than the template and process. The template can be written later based on some concrete examples.

I think all of the API docs belong in the user-facing docs and we shouldn't duplicate it here. Corner cases belong in test cases, with proper code comments. I think the design docs should just refer to these other places instead of duplicating the information.

Some information in the old RFCs, like rejected ideas, is only relevant before making a decision about a feature. That information belongs in issues and PR comments and not in the design docs.

If we add a lot of boilerplate (a verbose template encourages people to do that - a lot of folks think they're not allowed to delete anything so they fill in nonsense stuff instead or leave blank but keep the headings) then I believe these documents will be unmaintained very soon, maybe even before we've started using them. We already struggle keeping the user-facing docs up to date. Our docs need a lot of attention that they don't get.

**Comment by murphyjacob4** [comment]
> Why not start with the ASM design doc?

Added it now. Also removed the TEMPLATE.md and updated README.md to stress some of the points you called out. LMK what you think @zuiderkwast

**Review by zuiderkwast — APPROVED** [review]
You made it exactly to my liking! 😍  I hope you agree as well and didn't just change it for me. 😆

If we find out we want a template later, we can add it. I just prefer minimalism.

**Review by hpatro — APPROVED** [review]

**Comment by hpatro** [comment]
I was thinking we should kickoff this with IO threads redesign https://github.com/valkey-io/valkey/pull/3324. Nice to see @murphyjacob4 has already some design for atomic slot migration.

@valkey-io/core-team Are you folks good with this? We can request this for other new features going forward.

**Review by enjoy-binbin — APPROVED** [review]

**Comment by murphyjacob4** [comment]
Looks like we have 4/9 consensus. @madolson @ranshid @soloestoy @lucasyonge @PingXie  - any opposition to making the design docs folder? Otherwise will submit after 2 weeks from Hari's message

**Comment by murphyjacob4** [comment]
Looks like we 5/9 votes now. So we should be good to submit

**Comment by madolson** [comment]
@murphyjacob4 I'm slow, but I think this is fine. I think what to include is a bit weird, because I originally thought it was for inflight large designs but it seems like we're also using it for finalized smaller things.
